proposals/0000-clarify-execution-non-actor-async.md

```diff
@@ -0,0 +1,224 @@
+
+- It significantly improves the overhang problem relative to the current execution semantics.  Overhang would be bounded by the end of the call to an actor function, since upon return the caller would resume its own executor.
+
+- It is the only alternative which would make the current sendability rule for calls to `async` functions correct.
+
+However, it has three significant drawbacks:
+
+- While the overhang would be bounded, it would still cover potentially a large amount of code.  Everything called by an actor-isolated async function would resume to the actor, which could include a large amount of work that really doesn't need to be actor-isolated.  Actors could become heavily contended for artificial and perhaps surprising reasons.
+
+- It would make it difficult to write code that does leave the actor, since the inheritance would be implicit and recursive.  There could be an attribute which avoids the inheritance, but programmers would have to explicitly remember to use it.  This is the opposite of Swift's usual language design approach (e.g. with `mutating` methods); it's better to be less permissive by default so that the places which need stronger guarantees are explicit about it.
+
+- It would substantially impede optimization.  Since the current executor would be semantically observable by inheritance, optimizations that remove executor switches would still have to dynamically record the correct executor that should be inherited.  Since they currently do not do this, and since there is no efficient support in the runtime for doing this, this would come at a substantial runtime cost.
+
+### Initial inheritance of the caller's executor
+
+Another alternative would be to only inherit the executor of the caller for the initial period of execution, from the call to the first suspension.  Later resumptions would resume to a generic, non-actor executor.
+
+This would permit the current sendability rule for arguments, but only if we enforce that the parameters are not used after a suspension in the callee.  This is more flexible, but in ways that are highly likely to prove extremely brittle and limiting; a programmer relying on this flexibility is likely to come to regret it.  It would also still not permit return values to be non-`Sendable`, so the rule would still need changing.
+
+The overhang problem would be further improved relative to full inheritance.  The only real overhang risk would be a function that does a lot of synchronous work before returning or suspending.
+
+Sophisticated programmers might be able to use these semantics to avoid some needless switching.  It is common for `async` functions to begin with an `async` call, but if Swift has trouble analyzing the code that sets up that call, then under the proposed semantics, Swift might be unable to avoid the initial switch.  However, this optimization deficiency equally affects actor-isolated `async` functions, and arguably it ought to have a consistent solution.
```

Comment:
> Sophisticated programmers might be able to use these semantics to avoid some needless switching.

So this is definitely a cool trick but I agree this is more of something to _opt into_ and this way we can opt functions into it perhaps some day when "we know what we're doing and I promise not to block here" etc. Agreed on holistic solution, it's the same as actor-hop elimination of `func a() { await other.b()` :)

Reply:
Yeah, there's got to be a better way than this kind of subtle reasoning.